src/app/CommandSender.h

```diff
@@ -181,29 +181,98 @@ class CommandSender final : public Messaging::ExchangeDelegate
-        // set by `PrepareCommand` and is expected to be unchanged by caller until it is provided
-        // to `FinishCommand`.
+        PrepareCommandParameters & SetCommandRef(uint16_t aCommandRef)
+        {
+            commandRef.SetValue(aCommandRef);
+            return *this;
+        }
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response. We validate the reference
+        // early in PrepareCommand, even though it's not used until FinishCommand. This proactive
+        // validation helps prevent unnecessary writing an InvokeRequest into the packet that later
+        // needs to be undone.
+        //
+        // Currently, provided commandRefs for the first request must start at 0 and increment by one
+        // for each subsequent request. This requirement can be relaxed in the future if a compelling
+        // need arises.
         Optional<uint16_t> commandRef;
-        // For both `PrepareCommand` and `FinishCommand` this is an in parameter. It must have
-        // the same value when calling `PrepareCommand` and `FinishCommand` for a given command.
-        bool startOrEndDataStruct = false;
+        // If the InvokeRequest needs to be in a state with a started data TLV struct container
+        bool startDataStruct = false;
+    };
+
+    // FinishCommand is a public SDK API, so we must maintain source compatibility.
+    // Using this struct for API parameters instead of individual parameters allows us
+    // to make necessary changes for new functionality in a single location.
+    struct FinishCommandParameters
+    {
+        // gcc bug requires us to have the constructor below
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96645
+        FinishCommandParameters() {}
+
+        FinishCommandParameters(const Optional<uint16_t> & aTimedInvokeTimeoutMs) : timedInvokeTimeoutMs(aTimedInvokeTimeoutMs) {}
+        FinishCommandParameters(const AddRequestDataParameters & aAddRequestDataParam) :
+            timedInvokeTimeoutMs(aAddRequestDataParam.timedInvokeTimeoutMs), commandRef(aAddRequestDataParam.commandRef)
+        {}
+
+        FinishCommandParameters & SetEndDataStruct(bool aEndDataStruct)
+        {
+            endDataStruct = aEndDataStruct;
+            return *this;
+        }
+
+        FinishCommandParameters & SetCommandRef(uint16_t aCommandRef)
+        {
+            commandRef.SetValue(aCommandRef);
+            return *this;
+        }
+
+        // When a value is provided for timedInvokeTimeoutMs, this invoke becomes a timed
+        // invoke. CommandSender will use the minimum of all provided timeouts for execution.
+        const Optional<uint16_t> timedInvokeTimeoutMs;
+        // The command reference is required when sending multiple commands. It allows the caller
+        // to associate this request with its corresponding response.
+        Optional<uint16_t> commandRef;
```

Comment:
This needs to match the command ref provided in PrepareCommandParameters, yes?

Reply:
This is currently verified in `CommandSender::FinishCommand`. I have added a comment here to make that clear